src/Microsoft.Diagnostics.Tools.RuntimeClient/DiagnosticHelpers.cs

```diff
@@ -45,12 +45,78 @@ public static int GenerateCoreDump(int processId, string dumpName, DumpType dump
 
             byte[] serializedConfiguration;
             using (var stream = new MemoryStream())
-                serializedConfiguration = Serialize(header, stream, dumpName, dumpType, diagnostics);
+                serializedConfiguration = SerializeCoreDump(header, stream, dumpName, dumpType, diagnostics);
 
             return (int)EventPipeClient.SendCommand(processId, serializedConfiguration);
         }
 
-        private static byte[] Serialize(MessageHeader header, Stream stream, string dumpName, DumpType dumpType, bool diagnostics)
+        /// <summary>
+        /// Attach a profiler to the target process runtime.
+        /// </summary>
+        /// <param name="processId">.NET Core process id</param>
+        /// <param name="attachTimeout">The timeout (in ms) to wait while attempting to attach.</param>
+        /// <param name="profilerGuid">CLSID of the profiler to load</param>
+        /// <param name="profilerPath">Path to the profiler library on disk</param>
+        /// <param name="additionalData">additional data to pass to the profiler on attach</param>
+        /// <returns>HRESULT</returns>
+        public static int AttachProfiler(int processId, uint attachTimeout, Guid profilerGuid, string profilerPath, byte[] additionalData)
+        {
+            if (profilerGuid == null || profilerGuid == Guid.Empty)
+            {
+                throw new ArgumentException($"{nameof(profilerGuid)} must be a valid Guid");
+            }
+
+            if (String.IsNullOrEmpty(profilerPath))
+            {
+                throw new ArgumentException($"{nameof(profilerPath)} must be non-null");
+            }
+
+            var header = new MessageHeader {
+                RequestType = DiagnosticMessageType.AttachProfiler,
+                Pid = (uint)Process.GetCurrentProcess().Id,
+            };
+
+            byte[] serializedConfiguration;
+            using (var stream = new MemoryStream())
+            {
+                serializedConfiguration = SerializeProfilerAttach(header, stream, attachTimeout, profilerGuid, profilerPath, additionalData);
+            }
+
+            return (int)EventPipeClient.SendCommand(processId, serializedConfiguration);
```

Comment:
If the runtime client can't open a connection it will handle it gracefully and time out after a short while (maybe 5-10 seconds?) I ran in to it a couple times while testing when I accidentally used the wrong process ID.

I was thinking the attach timeout was just for the runtime when it tries to load the profiler, and the timeout for the runtime client trying to open a pipe should be a separate thing that is configurable for all requests (core dump, profiler attach, etc) independent of the profiler attach timeout.

Reply:
I think the attach timeout's original purpose was to be a timeout on the end-to-end action which included transmitting the request, acting on it, and then receiving some confirmation that it occured. Its true that we may want to add a general notion of a configurable timeout to all the APIs here, but so far profiler was the only one that had a timeout included as part of its pre-existing API. I suggested filing it as a todo so that if we do handle configurable timeouts in a general purpose way (which is likely) you may not need to do any extra work, I just didn't want to forget about it. I doubt users of the API will want to reason about multiple distinct timeouts for different portions of the operation.